### ISSUE TITLE
Revdep issues due to recursive proxying PR

### ISSUE THREAD

**Issue by lionel-** [r-lib/vctrs#1724]
Investigating failures in today's round of revdep I found this:

```r
df <- data_frame(x = ggplot2:::new_mapped_discrete(1:2))
list_unchop(list(df, df), list(1:2, 3:4))
#>   x
#> 1 1
#> 2 2
#> 3 1
#> 4 2
```

Became this after #1648:

```r
list_unchop(list(df, df), list(1:2, 3:4))
#>    x
#> 1 NA
#> 2 NA
#> 3 NA
#> 4 NA
```

Also:

```r
list_unchop(list(df, df))
#> Error in `list_unchop()`:
#> ! `proxy` of type `double` incompatible with `value` proxy of type `integer`.
#> ℹ In file slice-assign.c at line 162.
#> ℹ This is an internal error that was detected in the vctrs package.
#>   Please report it at <https://github.com/r-lib/vctrs/issues> with a reprex (<https://https://tidyverse.org/help/>) and the full backtrace.

vec_rbind(df, df)
#>   x
#> 1 1
#> 2 2
#> 3 1
#> 4 2
```

**Comment by DavisVaughan** [comment]
This problem happens with CRAN vctrs too:

```r
`proxy` of type `double` incompatible with `value` proxy of type `integer`.
```

Seems to come from `vec_assign_fallback()`, and can be attributed to this ggplot2 behavior with `[<-`

``` r
x <- ggplot2:::new_mapped_discrete(1:2)
typeof(x)
#> [1] "integer"

x[1L] <- x[1L]
typeof(x)
#> [1] "double"
```

From this method:

```r
> ggplot2:::`[<-.mapped_discrete`
function (x, ..., value) 
{
    if (length(value) == 0) {
        return(x)
    }
    value <- as.numeric(unclass(value))
    new_mapped_discrete(NextMethod())
}
```

This means that `[<-` here is type unstable. `x` started as an integer vector, but because `value <- as.numeric(unclass(value))` was run, the assignment done by base R in `NextMethod()` promoted `x` to a double vector.

I think the fact that this class uses a class list of `c("mapped_discrete", "numeric")` is going to be particularly painful. It would be easier if it just picked integer or double (likely double)

**Comment by thomasp85** [comment]
Is this based on the main branch? The release candidate contains a roll-back of mapped_discrete so it no longer builds on vctrs

**Comment by lionel-** [comment]
This is with CRAN ggplot2.

**Comment by thomasp85** [comment]
oh - then this is fixed in the next version already

**Comment by thomasp85** [comment]
However the class remains `c("mapped_discrete", "numeric")` because it is imperative that it inherit all the methods from numeric

**Comment by lionel-** [comment]
I confirmed this is fixed with `v3.4.0-rc`. Constructing a mapped-discrete with an integer vector is now an error, and with double inputs we get:

```r
df <- data_frame(x = ggplot2:::new_mapped_discrete(as.double(1:2)))

# Case 1
list_unchop(list(df, df), indices = list(1:2, 3:4))
#>   x
#> 1 1
#> 2 2
#> 3 1
#> 4 2

# Case 2
list_unchop(list(df, df))
#>   x
#> 1 1
#> 2 2
#> 3 1
#> 4 2
```

With double inputs, case 1 still fails with CRAN ggplot2. I think this needs some more investigation to make sure something isn't wrong with vctrs.

**Comment by thomasp85** [comment]
Be aware that with the next version ggplot2 switches to using the `mapped_discrete()`/`new_mapped_discrete()` constructor scheme advocated by vctrs
